### PR TITLE
[All] Remove use of compatibility layer 

### PR DESCRIPTION
**Pull request by fredroy**
⚠! Uber PR incoming! ⚠
(Will basically conflict every time a branch is merged 🤪) 

This PR generalizes the use of the Sofa.Type module (with things formerly in SofaHelper, SofaDefaulttype and SofaCore).
E.g:
 - helper::fixed_array becomes type::fixed_array
 - defaulttype::Vec becomes type::Vec
 - defaulttype::Quat and helper::Quater becomes type::Quat (and typedef *Quaternion* is removed)
 - helper::types::RGBAColor becomes type::RGBAColor
 - and many more
More info in #1521 

These renames was handled in Sofa.Compat to be the less problematic possible while converting.
Now for the release of v21.06, it is logical to use the correct namespaces/headers.

Big Warning though: the Sofa.Compat is still here for the external plugins, but as the headers/file inclusions changes, it is highly probable that they will break at compilation: a lot of people does not "include what they use" and what relying on the previous inclusions.

Later TODO: add a cmake option to make the use of Sofa.Compat optional.

______________________________________________________

By submitting this pull request, I acknowledge that  
**I have read, understand, and agree [SOFA Developer Certificate of Origin (DCO)](https://github.com/sofa-framework/sofa/blob/master/CONTRIBUTING.md#sofa-developer-certificate-of-origin-dco)**.
______________________________________________________

**Reviewers will merge this pull-request only if**  
- it builds with SUCCESS for all platforms on the CI.
- it does not generate new warnings.
- it does not generate new unit test failures.
- it does not generate new scene test failures.
- it does not break API compatibility.
- it is more than 1 week old (or has fast-merge label).

### CODE REVIEW

**Comment by fredroy** [comment]
[ci-build]

**Comment by fredroy** [comment]
[ci-build]

**Comment by fredroy** [comment]
[ci-build]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

**Comment by fredroy** [comment]
[ci-build][with-all-tests][force-full-build]

File: SofaKernel/modules/Sofa.GL/src/sofa/gl/BasicShapes.h [hunk 1]
```diff
@@ -175,7 +175,7 @@ void drawTorus(const float* coordinateMatrix, const float& bodyRad=0.0,  const f
 
 template <typename V>
 void drawEmptyParallelepiped(const V& vert1, const V& vert2, const V& vert3, const V& vert4, const V& vecFromFaceToOppositeFace, const float& rad=1.0, const int precision=8,
-                             const V& color=sofa::helper::fixed_array<int, 3>(255,0,0))
+                             const V& color=sofa::type::fixed_array<int, 3>(255,0,0))
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.GL/src/sofa/gl/BasicShapes.h]:
```suggestion
                             const V& color = sofa::type::RGBAColor::red()
```

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp [hunk 2]
```diff
@@ -330,8 +332,8 @@ TEST(QuaterTest, QuaterdOperatorDivRealSelf)
 
 //TEST(QuaterTest, QuaterdQuatVecMult)
 //{
-//    Quater<double> quat, quatMult;
-//    sofa::defaulttype::Vec<3, double> vec(1,2,3);
+//    Quat<double> quat, quatMult;
+//    sofa::type::Vec<3, double> vec(1,2,3);
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:336]: Should we keep commented code?

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp [hunk 3]
```diff
@@ -340,8 +342,8 @@ TEST(QuaterTest, QuaterdOperatorDivRealSelf)
 //
 //TEST(QuaterTest, QuaterdVecQuatMult)
 //{
-//    Quater<double> quat, quatMult;
-//    sofa::defaulttype::Vec<3, double> vec(1, 2, 3);
+//    Quat<double> quat, quatMult;
+//    sofa::type::Vec<3, double> vec(1, 2, 3);
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:346]: Should we keep this commented code?

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp [hunk 4]
```diff
@@ -517,10 +519,10 @@ TEST(QuaterTest, QuaterdCreateQuaterFromFrame)
 //TEST(QuaterTest, QuaterdCreateFromRotationVector)
 //{
 //    //axang2quat([1 2 5 pi / 12])
-//    sofa::defaulttype::Vec3d axis(1, 2, 5);
+//    sofa::type::Vec3d axis(1, 2, 5);
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:522]: Should we keep commented code?

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp [hunk 5]
```diff
@@ -544,10 +546,10 @@ TEST(QuaterTest, QuaterdCreateQuaterFromEuler)
 //TEST(QuaterTest, QuaterdCreateFromRotationVectorT)
 //{
 //    //axang2quat([1 2 5 pi / 12])
-//    sofa::defaulttype::Vec3d axis(1, 2, 5);
+//    sofa::type::Vec3d axis(1, 2, 5);
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:549]: Should we keep commented code?

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp [hunk 6]
```diff
@@ -574,8 +576,8 @@ TEST(QuaterTest, QuaterdQuatDiff)
 
 //TEST(QuaterTest, QuaterdAngularDisplacement)
 //{
-//    Quater<double> quat1, quat2;
-//    sofa::defaulttype::Vec3d displacement;
+//    Quat<double> quat1, quat2;
```
Comment (alxbilger) [inline SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:579]: Should we keep commented code?

File: SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/IntrUtility3.cpp [hunk 7]
```diff
@@ -25,6 +25,7 @@
 namespace sofa::component::collision
 {
 
+using namespace sofa::type;
 using namespace sofa::defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/IntrUtility3.cpp]: Is this one still useful?

File: SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/SphereModel.cpp [hunk 8]
```diff
@@ -26,6 +26,7 @@
 namespace sofa::component::collision
 {
 
+using namespace sofa::type;
 using namespace sofa::defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/SphereModel.cpp:30]: Is it still useful?

File: SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/SphereModel.inl [hunk 9]
```diff
@@ -120,6 +120,7 @@ void SphereCollisionModel<DataTypes>::draw(const core::visual::VisualParams* vpa
     if(d_componentState.getValue() != ComponentState::Valid)
         return ;
 
+    using namespace sofa::type;
     using namespace sofa::defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseCollision/src/SofaBaseCollision/SphereModel.inl:124]: Is it still useful?

File: SofaKernel/modules/SofaBaseLinearSolver/src/SofaBaseLinearSolver/matrix_bloc_traits.h [hunk 10]
```diff
@@ -120,39 +120,39 @@ class matrix_bloc_traits < defaulttype::Mat<L,C,real>, IndexType>
...
+//template<> inline const char* matrix_bloc_traits<type::Mat<4,4,float >, Size >::Name() { return "4f"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<4,4,double>, Size >::Name() { return "4d"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<6,6,float >, Size >::Name() { return "6f"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<6,6,double>, Size >::Name() { return "6d"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<8,8,float >, Size >::Name() { return "8f"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<8,8,double>, Size >::Name() { return "8d"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<9,9,float >, Size >::Name() { return "9f"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<9,9,double>, Size >::Name() { return "9d"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<12,12,float >, Size >::Name() { return "12f"; }
+//template<> inline const char* matrix_bloc_traits<type::Mat<12,12,double>, Size >::Name() { return "12d"; }
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseLinearSolver/src/SofaBaseLinearSolver/matrix_bloc_traits.h:155]: What do we do with all this commented code?

File: SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/DiagonalMass.cpp [hunk 11]
```diff
@@ -33,6 +33,7 @@ namespace sofa::component::mass
 
 using sofa::core::objectmodel::ComponentState ;
 
+using namespace sofa::type;
 using namespace sofa::defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/DiagonalMass.cpp:37]: Is `using namespace sofa::defaulttype;` still useful?

File: SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/UniformMass.cpp [hunk 12]
```diff
@@ -34,6 +34,7 @@ using std::string ;
 using std::ostringstream ;
 using sofa::component::mass::Vec3d ;
 using sofa::helper::system::DataRepository ;
+using namespace sofa::type;
 using namespace sofa::defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/UniformMass.cpp:38]: Is `using namespace sofa::defaulttype;` still useful?

File: SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TetrahedronNumericalIntegration_test.cpp [hunk 13]
```diff
@@ -36,6 +36,7 @@ using sofa::testing::NumericTest;
 namespace sofa {
 
 using namespace component;
+using namespace type;
 using namespace defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TetrahedronNumericalIntegration_test.cpp:40]: Is `using namespace sofa::defaulttype;` still useful?

File: SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TriangleNumericalIntegration_test.cpp [hunk 14]
```diff
@@ -37,6 +37,7 @@ using sofa::testing::NumericTest;
 namespace sofa {
 
 using namespace component;
+using namespace type;
 using namespace defaulttype;
```
Comment (alxbilger) [inline SofaKernel/modules/SofaBaseTopology/SofaBaseTopology_test/TriangleNumericalIntegration_test.cpp:41]: Is `using namespace sofa::defaulttype;` still useful?

**Review by alxbilger — CHANGES_REQUESTED** [review]

File: SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/DiagonalMass.cpp (hunk 11 above)
Comment (fredroy) [reply SofaKernel/modules/SofaBaseMechanics/src/SofaBaseMechanics/DiagonalMass.cpp:37]: Actually in that particular file yes (look at the explicite instanciations with RigidTypes, etc) but  there are a lot of files where it wont be useful for sure (but I did not really want to check for the bazillon number of  files)

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp (hunk 3 above)
Comment (fredroy) [reply SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:346]: obviously not but I wont clean the files 😴 in this PR

File: SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp (hunk 2 above)
Comment (damienmarchal) [reply SofaKernel/modules/Sofa.Type/Sofa.Type_test/Quater_test.cpp:336]: No never.

**Pull request merged**